### ISSUE TITLE
Deserialization of strings fails

### ISSUE THREAD

**Issue by funbiscuit** [SergioBenitez/ubyte#7]
Can't use serde deserialize feature because in `ser_de.rs` deserialization expects a u64:
```rust
impl<'de> Deserialize<'de> for ByteUnit {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where D: serde::Deserializer<'de>
    {
        deserializer.deserialize_u64(Visitor)
    }
}
```

Even though Visitor declares `visit_str`, it is not called because serde expects u64 and returns error when string is used.

**Comment by SergioBenitez** [comment]
Calling `deserialize_u64` merely serves as a hint for deserializers; it doesn't prohibit the deserializer from trying to deserialize a string. See the test cases which tests for string deserialization:

https://github.com/SergioBenitez/ubyte/blob/5f8b590dd3f5f4e026f1d2c553c08b329a1ca22b/src/ser_de.rs#L60-L83

See https://serde.rs/impl-deserialize.html and https://serde.rs/impl-deserializer.html for more.

**Comment by funbiscuit** [comment]
In my case it does prohibit deserializing. I've created a sample repository, please take a look:
https://github.com/funbiscuit/ubyte-deserialize-bug

Running it gives a following error:

/home/funbiscuit/.cargo/bin/cargo run --color=always --package deserialize-test --bin deserialize-test
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/deserialize-test`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("test: invalid type: string \"10MiB\", expected a byte unit as an integer or string", line: 1, column: 7)', src/main.rs:16:14
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: core::result::unwrap_failed
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/result.rs:1785:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/result.rs:1078:23
   4: deserialize_test::main
             at ./src/main.rs:15:9
   5: core::ops::function::FnOnce::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

Process finished with exit code 101

**Comment by lostl1ght** [comment]
Is this an issue with `serde_yaml 0.9` then? `0.8` works.

**Comment by koivunej** [comment]
Observing the same with `serde_json` v1.0.94 and `serde` v1.0.158.

By changing the `deserialize_u64` to `deserialize_any` it works. Looking at the `serde-json` impl, it seems that calls to `deserialize_u64` calls are [forwarded (with macro)](https://github.com/serde-rs/json/blob/4b9699612f57512b6734b43238c2382186245b31/src/de.rs#L1421) to [`deserialize_number`](https://github.com/serde-rs/json/blob/4b9699612f57512b6734b43238c2382186245b31/src/de.rs#L312) which errors out when it sees a non-number: https://github.com/serde-rs/json/blob/4b9699612f57512b6734b43238c2382186245b31/src/de.rs#L329.

> Calling deserialize_u64 merely serves as a hint for deserializers

Agreed that the impl-deserialize.html document suggests this, but it will not work in the case of json for example (or yaml -- didn't check but that was the original report). Unsure how this gap should be bridged since serde-json doesn't regard that as a hint but a requirement. Perhaps the answer is https://docs.rs/serde/latest/serde/trait.Deserializer.html#method.is_human_readable? Further it seems that `serde_test`'s `assert_de_tokens` forwards all `deserialize_*` methods to `deserialize_any` so it ends up hiding this.

I can PR this; use of `is_human_readable` causes some additional churn for the tests.

**Comment by reivilibre** [comment]
I also hit this and get errors like `invalid type: string "32B", expected a byte unit as an integer or string` or `invalid type: string "32 B", expected a byte unit as an integer or string` with `serde_json v1.0.107`.

**Comment by SergioBenitez** [comment]
Should be fixed in 0.10.4.
